WalletWasabi.Gui/Tabs/SettingsViewModel.cs

```diff
@@ -361,27 +312,42 @@ public string ValidateDustThreshold()
 			return "Invalid dust amount.";
 		}
 
-		public string ValidateHost(string host)
+		public string ValidateHost(string host, bool p2p)
 		{
 			if (string.IsNullOrWhiteSpace(host))
 			{
 				return string.Empty;
 			}
-
-			var theHost = host.Trim();
-			if (Uri.TryCreate(theHost, UriKind.Absolute, out _))
-			{
-				return string.Empty;
-			}
-			if (IPAddress.TryParse(theHost, out var ip))
+			try
 			{
-				if (ip.AddressFamily == AddressFamily.InterNetworkV6 && !Socket.OSSupportsIPv6)
+				if (p2p && !Config.TryNormalizeP2PHost(host, out host))
+				{
+					return "Invalid host";
+				}
+				
+				var endpoint = NBitcoin.Utils.ParseEndpoint(host, 9999);
+				if (endpoint.IsTor() && endpoint is DnsEndPoint dnsEndPoint)
+				{
+					try
+					{
+						var dnsHost = dnsEndPoint.Host;
+						NBitcoin.DataEncoders.Encoders.Base32.DecodeData(dnsHost.Remove(dnsHost.Length - ".onion".Length));
+					}
+					catch
+					{
+						return "Invalid onion host.";
+					}
+				}
+				if (endpoint is IPEndPoint ip &&
+					ip.Address.AddressFamily == AddressFamily.InterNetworkV6 &&
+					!endpoint.IsTor() && // Onioncat address would work even if the OS does not support IPv6
+					!Socket.OSSupportsIPv6)
 				{
 					return "OS does not support IPv6 addresses.";
 				}
 				return string.Empty;
 			}
-
+			catch { }
```

Comment:
Why is this?

Reply:
`ParseEndpoint` will throw if invalid host.